chainer_compiler/elichika/typing/ext/common.py

```diff
@@ -0,0 +1,78 @@
+from   chainer_compiler.elichika.typing.ext.utils  import *
+from   chainer_compiler.elichika.typing.types      import *
+from   chainer_compiler.elichika.typing.shape_elem import *
+
+__all__ = [ 'ty_MakeTensor'
+          , 'ty_TensorArith'
+          ]
+
+
+class ty_MakeTensor():
+    def calculate_shape(self, x_type):
+        if not isinstance(x_type, TySequence):
+            return ()
+        if not x_type.is_fixed_len:
+            return (None,)
+
+        x_tys = x_type.get_tys()
+
+        if isinstance(x_tys[0], TySequence):
+            list_lengths = [t.size() if t.is_fixed_len else None for t in x_tys]
+            list_lengths_nonnull = [l for l in list_lengths if l is not None]
+
+            # for example, we will not accept np.array([[1,2,3], [4,5]])
+            assert all_same(list_lengths_nonnull), \
+                    "incompatible list length"
+
+        return (len(x_tys),) + self.calculate_shape(x_tys[0])
+
+    def get_element_dtype(self, ty):
+        # get element dtype of nested TySequence
+        if isinstance(ty, TySequence):
+            return self.get_element_dtype(ty.get())
+        return tyobj_to_dtype(ty)
+
+
+
+class ty_TensorArith():
+    def __init__(self, kind):
+        self.kind = kind
+
+    def __call__(self, ty_args, ty_kwargs):
+        x_type, y_type = ty_args
+        x_shape = x_type.shape if isinstance(x_type, TyTensor) else ()
+        y_shape = y_type.shape if isinstance(y_type, TyTensor) else ()
+        dtype = self.get_out_dtype(x_type, y_type)
+
+        if len(x_shape) > len(y_shape):
+            shape = self.infer_return_shape(x_shape, y_shape)
+        else:
+            shape = self.infer_return_shape(y_shape, x_shape)
+        return TyTensor(self.kind, dtype, shape)
+
+    def infer_return_shape(self, x_shape, y_shape):
+        ret_shape = [None for _ in x_shape]
+        for i in range(len(x_shape) - len(y_shape)):
+            ret_shape[i] = copy_ShapeElem(x_shape[i])
+        for i in range(1, len(y_shape) + 1):
```

Comment:
Probably, `reversed(range(len(y_shape)))` (and s/-i/i/g) would be easier to read?

Reply:
Thanks! I think it's equivalent to `range(len(x_shape) - len(y_shape), len(x_shape))` and s/`y_shape[i]`/`y_shape[i - len(x_shape) + len(y_shape)]`/g .

Reply:
Ah, I read code wrongly. I thought the size of `y_shape` is equal to `x_shape`/`ret_shape`. Then, I think the original code looks slightly better to me, but the current code would also make sense. Sorry for the back-and-forth. It's up to you.

Reply:
I prefer the original version, too. I'll revert the commit. Thanks!